ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Offset.scala

```diff
@@ -4,46 +4,40 @@
 package com.daml.ledger.participant.state.v1
 
 import com.digitalasset.daml.lf.data.Ref.LedgerString
-import com.digitalasset.daml.lf.data.Ref
+
+import scala.util.Try
 
 /** Offsets into streams with hierarchical addressing.
   *
   * We use these [[Offset]]'s to address changes to the participant state.
-  * We allow for array of [[Long]] to allow for hierarchical addresses.
-  * These [[Long]] values are expected to be positive. Offsets are ordered by
-  * lexicographic ordering of the array elements.
+  * Offsets are arbitrary string values. Offset strings must be strict
+  * monotonically increasing according to lexicographical ordering.
   *
-  * A typical use case for [[Offset]]s would be addressing a transaction in a
-  * blockchain by `[<blockheight>, <transactionId>]`. Depending on the
-  * structure of the underlying ledger these offsets are more or less
-  * nested, which is why we use an array of [[Long]]s. The expectation is
-  * though that there usually are few elements in the array.
+  * Ledger implementations are advised to future proof their design
+  * of offsets by reserving the first (few) characters to a version
+  * indicator, followed by the specific offset scheme for that version.
+  * This way it is possible in the future to switch to a different versioning
+  * scheme, while making sure that previously created offsets are always
+  * less than newer offsets.
   *
   */
-final case class Offset(private val xs: Array[Long]) extends Ordered[Offset] {
-  def toLedgerString: Ref.LedgerString =
-    // It is safe to concatenate number and "-" to obtain a valid transactionId
-    Ref.LedgerString.assertFromString(components.mkString("-"))
-
-  override def toString: String = toLedgerString
+final case class Offset private (value: String) extends Ordered[Offset] {
 
-  def components: Iterable[Long] = xs
+  def toLedgerString: LedgerString = LedgerString.assertFromString(toString)
 
-  override def equals(that: Any): Boolean = that match {
-    case o: Offset => this.compare(o) == 0
-    case _ => false
-  }
+  override def toString: String = value.toString()
 
-  def compare(that: Offset): Int =
-    scala.math.Ordering.Iterable[Long].compare(this.xs.toIterable, that.xs.toIterable)
+  override def compare(that: Offset): Int = value.compareTo(that.value)
 }
 
 object Offset {
 
-  /** Create an offset from a string of form 1-2-3. Throws
-    * NumberFormatException on misformatted strings.
-    */
-  def assertFromString(s: LedgerString): Offset =
-    Offset(s.split('-').map(_.toLong))
+  def empty: Offset = Offset("")
+
+  def assertFromString(s: String): Offset =
+    Offset(s)
+
+  def fromString(s: String): Try[Offset] = Try(assertFromString(s))
 
+  def fromLong(l: Long, padding: Int = 8): Offset = assertFromString(s"%0${padding}d".format(l))
```

Comment:
This is used in quite a few places. I'm tempted to removed it so that each ledger implementation has to make a conscious decision about the offsets. For KV Ledgers, kvutils could provide a small utility library to add further restrictions (e.g. kvutils needs to manipulate returned offsets to [account for multiple state updates per log entry](https://github.com/digital-asset/daml/blob/d333d9e8820770875876d2639127b531a2e4520c/ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/KeyValueParticipantStateReader.scala#L32-L35)).